Transformations: Add "Rename By Regex" transformer

simianhacker opened the pull request:
<!--

Thank you for sending a pull request! Here are some tips:

1. If this is your first time, please read our contribution guide at https://github.com/grafana/grafana/blob/master/CONTRIBUTING.md

2. Ensure you include and run the appropriate tests as part of your Pull Request.

3. In a new feature or configuration option, an update to the documentation is necessary. Everything related to the documentation is under the docs folder in the root of the repository.

4. If the Pull Request is a work in progress, make use of GitHub's "Draft PR" feature and mark it as such.

5. If you can not merge your Pull Request due to a merge conflict, Rebase it. This gets it in sync with the master branch.

6. Name your PR as "<FeatureArea>: Describe your change", e.g. Alerting: Prevent race condition. If it's a fix or feature relevant for the changelog describe the user impact in the title. The PR title is used to auto-generate the changelog for issues marked with the "add to changelog" label.

-->

**What this PR does / why we need it**:

This PR allows users to rename metrics by using a regular expression and replacement pattern via a transformer. This is useful for when you're doing a terms aggregation in something like Elasticsearch and you want to modify the values of the terms. For example, let's say you have a terms agg on `host.name` your team is using a pattern like `web-{number}.example.com` and you only want to display `web-{number}` in the results. You would set the `Match` field to `([^.]+).example.com` and the `Replace` field. to `$1`; `web-01.example.com` would be come `web-01`.

In the screenshot below, this transformer is removing the first part of the `event.dataset` and only displaying the suffix, `system.cpu` to `cpu`.

![image](https://user-images.githubusercontent.com/41702/99857631-dc624700-2b48-11eb-9a96-8d3e2661c9a4.png)

This was originally requested in #5212 and subsequently requested in #29133. @torkelo mentioned that this should be done as a transformer in #29133.


**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

Fixes #5212

**Special notes for your reviewer**:

Let me know if there is a better name then `Rename by regex` or if we need to change the form layout.

simianhacker committed:
Grafana: Rename By Regex Transformer

simianhacker committed:
Removing unused deps

simianhacker committed:
Merge branch 'master' of github.com:grafana/grafana into issue-5212-rename-by-regex

simianhacker committed:
Add scrollIntoView() to TranformTab.content()

simianhacker committed:
Exporting RenameByRegexTransformerOptions

simianhacker committed:
Merge branch 'master' of github.com:grafana/grafana into issue-5212-rename-by-regex

simianhacker committed:
Add doc block to renameByRegex transformer

simianhacker committed:
Adding doc block for RenameByRegexTransformerOptions

simianhacker committed:
removing scrollIntoView() for transform tab

simianhacker committed:
Adding back in scrollIntoView() for transform panel

simianhacker commented:
Is there a way to get access to the video output of the E2E tests on CI? These tests pass locally when I run `yarn e2e` but I'm unable to get the same tests to pass in CI. If you look at the commit history, I had to add `scrollIntoView()` for the transform panel to get it to pass locally but if I remove that, it fails.

hugohaggmark commented:
> Is there a way to get access to the video output of the E2E tests on CI? These tests pass locally when I run `yarn e2e` but I'm unable to get the same tests to pass in CI. If you look at the commit history, I had to add `scrollIntoView()` for the transform panel to get it to pass locally but if I remove that, it fails.

Hi @simianhacker we used to, but we've recently switched to Drone, so right now it isn't possible. I'll checkout your branch and have a look to see if I can figure out what is going on.
(ping @aknuds1)

hugohaggmark committed:
Tests: fixes e2e tests

hugohaggmark commented:
@simianhacker I pushed some changes to the e2e tests that should fix them. I hope that is Ok.

simianhacker commented:
Totally OK! Thank you for the help!

hugohaggmark reviewed (commented):
I think this PR is off to a brilliant start, but I hesitate to approve it because this transformer only operates on the name of frames. I think we need to incorporate renaming of field names and labels too in order to make it more intuitive for all use cases.

simianhacker commented:
@hugohaggmark Thanks for the feedback! I will apply it to fields and labels as well similar to `Filter by name`.

simianhacker committed:
Apply to displayName instead of just the name of the frame

simianhacker commented:
@hugohaggmark I think the changes I just added should work as you suggested. I read through the `getDisplayName()` code and it appears to handle the labels as well, let me know if I've misinterpreted how that works.

simianhacker committed:
Rewrite docblock to match new functionality

simianhacker committed:
Merge branch 'master' of github.com:grafana/grafana into issue-5212-rename-by-regex

mckn reviewed (changes requested):
I think this looks great! Well done.

However I am missing some documentation for this. It would be awesome if you can add some information about how to use this with some examples here: https://github.com/grafana/grafana/blob/master/docs/sources/panels/transformations/types-options.md

Also - I think adding support for selecting at what level of the data frame you want to replace the naming e.g. so you can select either field, frame or labels. But I think we can merge this PR and add support for the other types in follow up PRs.

simianhacker committed:
Merge branch 'master' of github.com:grafana/grafana into issue-5212-rename-by-regex

simianhacker commented:
> Also - I think adding support for selecting at what level of the data frame you want to replace the naming e.g. so you can select either field, frame or labels. But I think we can merge this PR and add support for the other types in follow up PRs.

@mckn In most of the other transformers, they seem to limit the functions to just the field, frame and labels without specifying which one to apply it to; are there plans to add that functionality to those transformers? 

@hugohaggmark What's the harm in applying this to the values as well? For example, one might want to modify the values returned from a term's aggregation in a data table, like below:

![image](https://user-images.githubusercontent.com/41702/100794123-62e51700-33da-11eb-9061-314888ff80d1.png)

to something like this

![image](https://user-images.githubusercontent.com/41702/100797050-c40ee980-33de-11eb-8412-ce473be65230.png)

Or should that be a seperate transformer? "Replace values by regex" maybe? I put together a quick function that could easily accomplish this as well (for the screenshot above).

```typescript
const replaceValuesByRegex = (options: RenameByRegexTransformerOptions) => (frame: DataFrame) => {
  const regex = new RegExp(options.regex);
  const fields = frame.fields.map(field => {
    if (field.type !== 'string') {
      return field;
    }
    const values = [];
    for (let index = 0; index < field.values.length; index++) {
      const value = field.values.get(index);
      if (regex.test(value)) {
        values.push(value.replace(regex, options.renamePattern));
      } else {
        values.push(value);
      }
    }
    return { ...field, values: new ArrayVector(values) };
  });
  return { ...frame, fields };
};
```

@mckn I started working on the documentation today but I wanted to resolve these questions before I finalize it.

hugohaggmark commented:
@simianhacker we've recently added a transformer that filters by value, so I think adding a new one that renames by value should be ok. But I would wait a bit with that one, maybe create a feature request to gather more feedback from the community. WDYT?

simianhacker commented:
@hugohaggmark Sounds good, I will get the docs done and work on wrapping this one up.

simianhacker committed:
Adding documentation

simianhacker committed:
Changing TLD to domain name

simianhacker commented:
@hugohaggmark I added some documentation, writing is not my forté so please feel free to suggest some edits. I also created some images for the documentation but I couldn't figure out where those images are stored in the repo.

simianhacker committed:
Merge branch 'master' of github.com:grafana/grafana into issue-5212-rename-by-regex

hugohaggmark commented:
> @hugohaggmark I added some documentation, writing is not my forté so please feel free to suggest some edits. I also created some images for the documentation but I couldn't figure out where those images are stored in the repo.

Maybe @oddlittlebird can help with the images as I really don't know how to either?

achatterjee-grafana commented:
@hugohaggmark, images are kept in the website repo. Diana, how will we add the images in? Thanks.

simianhacker committed:
Fixing typo

oddlittlebird commented:
The images are stored in a private repo. If you or Hugo can take some screenshots and then attach them here (please label them with what they are and where they should go, then a Grafana maintainer can add them to the private website repo in a separate PR.

oddlittlebird commented:
The addition to the docs looks good to me. Thanks for including examples!

simianhacker commented:
@oddlittlebird Here are the images

#### rename-by-regex-before.png

<img width="1367" alt="rename-by-regex-before" src="https://user-images.githubusercontent.com/41702/101092294-ad02ff80-3576-11eb-97d7-cc546649d0fb.png">

#### rename-by-regex-after.png

<img width="1367" alt="rename-by-regex-after" src="https://user-images.githubusercontent.com/41702/101092339-bf7d3900-3576-11eb-83a6-3edad224b255.png">

osg-grafana commented:
@oddlittlebird What are the guidelines for image size, width, height, and file type? I expect that we can address such questions outside of this particular PR, and this is the context in which I thought to ask this question.

oddlittlebird commented on docs/sources/panels/transformations/types-options.md:
```diff
@@ -336,3 +337,18 @@ When you have more than one condition, you can choose if you want the action (in
+
+With the transformation applied, you can see we are left with just the remainder of the string.
+
+{{< docs-imagebox img="/img/docs/transformations/rename-by-regex-after.png" class="docs-image--no-shadow" max-width= "1100px" >}}
```
```suggestion
{{< docs-imagebox img="/img/docs/transformations/rename-by-regex-after-7-3.png" class="docs-image--no-shadow" max-width= "1100px" >}}
```

oddlittlebird commented on docs/sources/panels/transformations/types-options.md:
```diff
@@ -336,3 +337,18 @@ When you have more than one condition, you can choose if you want the action (in
+
+Here is an example where we are stripping the prefix from event types, in the before you can see everything is prefixed with `system.`
+
+{{< docs-imagebox img="/img/docs/transformations/rename-by-regex-before.png" class="docs-image--no-shadow" max-width= "1100px" >}}
```
```suggestion
{{< docs-imagebox img="/img/docs/transformations/rename-by-regex-before-7-3.png" class="docs-image--no-shadow" max-width= "1100px" >}}
```

oddlittlebird reviewed (commented):
rename-by-regex-before-7-3.png and rename-by-regex-after-7.3.png have been merged to the website repo.

oddlittlebird commented:
> @oddlittlebird What are the guidelines for image size, width, height, and file type? I expect that we can address such questions outside of this particular PR, and this is the context in which I thought to ask this question.

Image guidelines are in the [Grafana documentation style guide](https://github.com/grafana/grafana/blob/master/contribute/style-guides/documentation-style-guide.md#images).

simianhacker committed:
Update docs/sources/panels/transformations/types-options.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

simianhacker committed:
Update docs/sources/panels/transformations/types-options.md

Co-authored-by: Diana Payton <52059945+oddlittlebird@users.noreply.github.com>

achatterjee-grafana commented on docs/sources/panels/transformations/types-options.md:
```diff
@@ -336,3 +337,17 @@ When you have more than one condition, you can choose if you want the action (in
+
+You can specify a regular expression, which is only applied to matches, along with a replacement pattern that support back references. For example, let's imagine you're visualizing CPU usage per host and you want to remove the domain name. You could set the regex to `([^\.]+)\..+` and the replacement pattern to `$1`, `web-01.example.com` would become `web-01`.
+
+Here is an example where we are stripping the prefix from event types, in the before you can see everything is prefixed with `system.`
```
```suggestion
In the following example, we are stripping the prefix from event types. In the before image, you can see everything is prefixed with `system.`
```

achatterjee-grafana commented on docs/sources/panels/transformations/types-options.md:
```diff
@@ -336,3 +337,17 @@ When you have more than one condition, you can choose if you want the action (in
+
+Here is an example where we are stripping the prefix from event types, in the before you can see everything is prefixed with `system.`
+
+{{< docs-imagebox img="/img/docs/transformations/rename-by-regex-before-7-3.png" class="docs-image--no-shadow" max-width= "1100px" >}}
```
Why are the images named before-7-3? If we are adding new images, should we capture them from the latest build?

achatterjee-grafana reviewed (commented):
Added a few comments.

simianhacker committed:
Update docs/sources/panels/transformations/types-options.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

mckn reviewed (approved):
Awesome job on this!

charlescva commented:
Woot! 4.5 years after asking for this in issue #5212 and it's here! Great job!

oddlittlebird commented:
@mckn , is this going to be in 7.4? If so, then we need to add a note to the docs indicating that immediately. It will cause user confusion otherwise.

mckn commented:
> @mckn , is this going to be in 7.4? If so, then we need to add a note to the docs indicating that immediately. It will cause user confusion otherwise.

Yes, added in this PR: #29735. Great catch!

dicer commented:
> @simianhacker we've recently added a transformer that filters by value, so I think adding a new one that renames by value should be ok. But I would wait a bit with that one, maybe create a feature request to gather more feedback from the community. WDYT?

Is there a feature request or PR for this yet? Would this allow me to rename all kinds of values? I have some queries that return strings like "closed" and I would like to rename them to "0" (and "open" to "1"), so I can graph them.

simianhacker commented:
> Is there a feature request or PR for this yet? Would this allow me to rename all kinds of values? I have some queries that return strings like "closed" and I would like to rename them to "0" (and "open" to "1"), so I can graph them.

@dicer There isn't a feature request or PR, I had added it as part of this change but then pulled it out per the conversation above. If we implemented this for values, it would give you all the power of a find/replace regular expression. You could create one transformer that replaced all the zeros with `closed` and a second one that replaces all the ones with `open`.

oddlittlebird commented:
Folks, please open a new issue rather than having discussions on merged PRs.
